Storage/api.h

```diff
@@ -440,22 +456,79 @@ class RESTfulStorage {
     }
   }
 
+  /*
+  RESTfulStorage(const RESTfulStorage&) = delete;
+  RESTfulStorage(RESTfulStorage&&) = default;
+  RESTfulStorage& operator=(const RESTfulStorage&) = delete;
+  RESTfulStorage& operator=(RESTfulStorage&&) = default;
+  */
+
+  template <class QUERY_IMPL>
+  void AddCQRSQuery(const std::string& query) {
+    std::lock_guard<std::mutex> lock(data_->cqrs_handlers_mutex_);
+    // TODO(dkorolev): Exception if the handler is already registered.
+    data_->cqrs_query_map_[query] = std::make_pair(
+        [](const Request& request) -> std::shared_ptr<CurrentSuper> { return ParseCQRSRequest<QUERY_IMPL>(request); },
+        [query](immutable_fields_t fields, std::shared_ptr<CurrentSuper> type_erased_query, const std::string& url)
+            -> Response {
+              return dynamic_cast<QUERY_IMPL&>(*type_erased_query.get())
+                  .template Query<ImmutableFields<STORAGE_IMPL>>(fields, url);
+            });
+  }
+
+  template <class COMMAND_IMPL>
+  void AddCQRSCommand(const std::string& command) {
+    std::lock_guard<std::mutex> lock(data_->cqrs_handlers_mutex_);
+    // TODO(dkorolev): Exception if the handler is already registered.
+    data_->cqrs_command_map_[command] = std::make_pair(
+        [](const Request& request) -> std::shared_ptr<CurrentSuper> { return ParseCQRSRequest<COMMAND_IMPL>(request); },
+        [command](mutable_fields_t fields, std::shared_ptr<CurrentSuper> type_erased_command, const std::string& url)
+            -> Response {
+              fields.SetTransactionMetaField("X-Current-CQRS-Command", command);
+              // TODO(dkorolev): What do we do on user code exceptions here?
+              try {
+                if (current::reflection::FieldCounter<COMMAND_IMPL>::value > 0u) {
+                  // return ParseJSON<COMMAND_IMPL>(http_body).
+                  return dynamic_cast<COMMAND_IMPL&>(*type_erased_command.get())
+                      .template Command<MutableFields<STORAGE_IMPL>>(fields, url);
+                } else {
+                  return COMMAND_IMPL().template Command<MutableFields<STORAGE_IMPL>>(fields, url);
+                }
+              } catch (const TypeSystemParseJSONException& e) {
+                return Response(cqrs::CQRSParseJSONException(e.What()), HTTPResponseCode.BadRequest);
+              }
+              // TODO(dkorolev): Test transaction rollbacks.
+              return Response(cqrs::CQRSBadRequest(), HTTPResponseCode.BadRequest);
```

Comment:
The code will get here if the `try` blocks throws an exception which is not `TypeSystemParseJSONException`, and not the one inherited from `TypeSystemParseJSONException`.